File: discrepancy.py
#!/usr/bin/python
import pandas as pd 
import numpy as np
import itertools as itt
import time 
import os 

# pre-processing
data = pd.read_csv('input.csv')
target = 210.04
data['minus'] = -data['minus']

arr = np.r_[np.array(data['plus']),np.array(data['minus'])]
arr = arr[arr!=0]
arr = np.unique(arr)

# finding combos 
def combine_any(num):

	# lazy eval, itertools object is a map of tuples
	comb = itt.combinations(arr, num)

	# create a pandas DataFrame
	df = pd.DataFrame(list(comb))
	df['sum'] = round(df.sum(axis = 1), 2)

	# target_exists = target in df['sum'] 
	# precision error comparing float and np.float64
	
	# create difference column
	df['diff'] = df['sum'] - target 

	# establish tolerance level, use bool to subset 
	tolerance = np.finfo(np.float64).eps
	is_basically_zero = abs(df['diff']) < tolerance
	x = df[is_basically_zero]

	# get path and filename
	path = os.getcwd()
	filename = 'combination'+str(num)+'.csv'
	
	# write out csv 
	x.to_csv(os.path.join(path, filename))


start = time.time()
for i in range(2, 6):
	combine_any(i)
end = time.time()

print(end - start)

# In Python:
# 25.197231769561768 secs

File: discrepancy.R

# PRE-PROCESSING

# read csv file 
rm(list=ls())
dfm <- read.csv('input.csv')

# get target value from user 
#target <- readline(prompt='Enter discrepancy amount: ')
target <- '210.04'

# change second column to negative
dfm$minus <- -dfm$minus

# combine both columns into a single vector
vec <- c(dfm$plus,dfm$minus)

# remove all zeroes and duplicates 
vec <- as.numeric(lapply(vec,function(x) x[x!=0]))
vec <- vec[!is.na(vec)]
vec <- sort(unique(vec))


# FIGURING OUT COMBINATIONS

combineAny <- function(num) {

	# get dataframe of combos
	comb <- as.data.frame(t(combn(vec, num)))
	
	# create Sum column 
	if (num == 2) {
		comb$Sum <- comb$V1+comb$V2
	} else if (num == 3) {
		comb$Sum <- comb$V1+comb$V2+comb$V3
	} else if (num == 4) {
		comb$Sum <- comb$V1+comb$V2+comb$V3+comb$V4
	} else {
		comb$Sum <- comb$V1+comb$V2+comb$V3+comb$V4+comb$V5
	}

 	# find whether target Sum exists  
	target_exists <- ifelse(target %in% comb$Sum, 1, 0)


	if (target_exists == 1) {

		# get dataframe of unique combos
		x <- unique(comb[comb$Sum == target,])
		x <- x[,-which(colnames(x) == 'Sum')]
		
		# sort columns and remove duplicate rows 
		for (i in 1:nrow(x)) {
			x[i,] <- sort(x[i,])
		}
		x <- unique(x)
		
		# write out csv with results 
		write.csv(x, paste('combo',num,'.csv',sep=''))
	}
}



system.time({ 

	# run combineAny() for combos of 2-to-5 figures
	for (i in 2:5) {
		combineAny(i)
	}
})

#   user  system elapsed 
#  73.66    0.90   75.30 
	


	
	
	
	
	
	
	
	
	
	
	




